libs/langchain-google-common/src/connection.ts

```diff
@@ -298,141 +382,216 @@ export abstract class AbstractGoogleLLMConnection<
-      topK: parameters.topK,
-      topP: parameters.topP,
-      maxOutputTokens: parameters.maxOutputTokens,
-      stopSequences: parameters.stopSequences,
-      responseMimeType: parameters.responseMimeType,
+      subEvent: names[1],
+      module: names[2],
     };
   }
 
-  formatSafetySettings(
-    _input: MessageType,
-    parameters: GoogleAIModelRequestParams
-  ): GeminiSafetySetting[] {
-    return parameters.safetySettings ?? [];
-  }
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  abstract handleCustomRequestEvent(
+    eventName: string,
+    eventInfo: GoogleCustomEventInfo,
+    data: any,
+    runId: string,
+    tags?: string[],
+    metadata?: Record<string, any>
+  ): any;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  abstract handleCustomResponseEvent(
+    eventName: string,
+    eventInfo: GoogleCustomEventInfo,
+    data: any,
+    runId: string,
+    tags?: string[],
+    metadata?: Record<string, any>
+  ): any;
 
-  async formatSystemInstruction(
-    _input: MessageType,
-    _parameters: GoogleAIModelRequestParams
-  ): Promise<GeminiContent> {
-    return {} as GeminiContent;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  abstract handleCustomChunkEvent(
+    eventName: string,
+    eventInfo: GoogleCustomEventInfo,
+    data: any,
+    runId: string,
+    tags?: string[],
+    metadata?: Record<string, any>
+  ): any;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  handleCustomEvent(
+    eventName: string,
+    data: any,
+    runId: string,
+    tags?: string[],
+    metadata?: Record<string, any>
+  ): any {
+    if (!eventName) {
+      return undefined;
+    }
+    const eventInfo = this.customEventInfo(eventName);
+    switch (eventInfo.subEvent) {
+      case "request":
+        return this.handleCustomRequestEvent(
+          eventName,
+          eventInfo,
+          data,
+          runId,
+          tags,
+          metadata
+        );
+      case "response":
+        return this.handleCustomResponseEvent(
+          eventName,
+          eventInfo,
+          data,
+          runId,
+          tags,
+          metadata
+        );
+      case "chunk":
+        return this.handleCustomChunkEvent(
+          eventName,
+          eventInfo,
+          data,
+          runId,
+          tags,
+          metadata
+        );
+      default:
+        console.error(
+          `Unexpected eventInfo for ${eventName} ${JSON.stringify(
+            eventInfo,
+            null,
+            1
+          )}`
+        );
+    }
   }
+}
 
-  structuredToolToFunctionDeclaration(
-    tool: StructuredToolParams
-  ): GeminiFunctionDeclaration {
-    const jsonSchema = zodToGeminiParameters(tool.schema);
-    return {
-      name: tool.name,
-      description: tool.description ?? `A function available to call.`,
-      parameters: jsonSchema,
-    };
+export class GoogleRequestLogger extends GoogleRequestCallbackHandler {
```

Comment:
This is intended for public use? Or just your own debugging?

Reply:
Initially for my debugging, but there have been cases where people have reported problems that I couldn't duplicate. Being able to easily have them add this logger to the callbacks to see what the request and response were would make it easier to create a test and a fix.

So it is rare that anyone would need to use it, but invaluable to have in place if they do.